### PR TITLE
bugfix: make msc3967 idempotent

### PR DESCRIPTION
**Pull request by kegsay**
MSC3967 was updated recently to make it more robust to network failures:

> there is an existing cross-signing master key and it exactly matches the cross-signing master key provided in the request body. If there are any additional keys provided in the request (self signing key, user signing key) they MUST also match the existing keys stored on the server. In other words, the request contains no new keys. If there are new keys, UIA MUST be performed.

https://github.com/matrix-org/matrix-spec-proposals/blob/hughns/device-signing-upload-uia/proposals/3967-device-signing-upload-uia.md#proposal

This covers the case where the 200 OK is lost in transit so the client retries the upload, only to then get UIA'd.

Complement tests: https://github.com/matrix-org/complement/pull/713 - passing example https://github.com/element-hq/synapse/actions/runs/7976948122/job/21778795094?pr=16943#step:7:8820

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Review by reivilibre — APPROVED** [review]

File: synapse/handlers/e2e_keys.py [hunk 1]
```diff
@@ -1476,6 +1476,23 @@ async def check_cross_signing_setup(self, user_id: str) -> Tuple[bool, bool]:
...
+    async def has_different_keys(self, user_id: str, body: JsonDict) -> bool:
+        # Ensure that each key provided in the request body exactly matches the one we have stored.
+        # The first time we see a difference, bail.
+        req_body_key_to_db_key = {
+            "master_key": "master",
+            "self_signing_key": "self_signing",
+            "user_signing_key": "user_signing",
+        }
+        for req_body_key, db_key in req_body_key_to_db_key.items():
+            if req_body_key in body:
```
[the remark this answers is not in the document]
Comment (kegsay) [reply synapse/handlers/e2e_keys.py:1488]: The body contains a set of keys that need to be updated in the DB (i.e data flows from the body to the DB). If the body doesn't have a key but the DB does, there is no work to do as there is no new key.

[the remark this answers is not in the document]
Comment (reivilibre) [reply synapse/handlers/e2e_keys.py:1488]: Ah, I see it now. I think this method could use a docstring update, maybe even a name change (something to do with `conflict`?).

Basically, just something to drive the point home that it's not allowed to overwrite an existing key with a different key (only with itself as an idempotence measure). But of course writing a new key when there wasn't one before is fine. I guess I got too hung up on the word 'difference' when we are more or less talking about a one-way check?

File: changelog.d/16943.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Make the CSAPI endpoint /keys/device_signing/upload idempotent.
```
Comment (reivilibre) [inline changelog.d/16943.bugfix:1]:
```suggestion
Make the CSAPI endpoint `/keys/device_signing/upload` idempotent.
```

**Review by reivilibre — APPROVED** [review]

File: synapse/handlers/e2e_keys.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kegsay) [reply synapse/handlers/e2e_keys.py:1488]: Ah I see your confusion with "difference", I'll add more words.

**Pull request merged**